src/passes/OptimizeInstructions.cpp

```diff
@@ -71,7 +72,16 @@ struct PatternDatabase {
   };
 };
 
-static PatternDatabase database;
+static PatternDatabase* database = nullptr;
+
+static void ensureDatabase() {
+  if (!database) {
+    // we must only ever create one database
+    static OnlyOnce onlyOnce;
+    onlyOnce.verify();
+    database = new PatternDatabase;
```

Comment:
Almost, yeah, but to thread-safely initialize the static variable once, it would take a lock, wouldn't it? This code avoids that (by relying on the fact that we'll reach this on the main thread first).